api/api-doc/column_family.json

```diff
@@ -92,6 +92,14 @@
                      "type":"boolean",
                      "paramType":"query"
                   },
+                  {
+                     "name":"aggressive_purge_tombstones",
```

Comment:
but I'm not opposed to this change or this option.
I actually think it can be enabled by default since major-compaction is an admin operation and the admin should be aware of what it is doing.

Reply:
The row level check can indeed greatly help, although it will significantly reduce the performance of compaction.
Note that it will not work for commitlog -- we cannot read from the commitlog (currently).

Reply:
> The row level check can indeed greatly help, although it will significantly reduce the performance of compaction. Note that it will not work for commitlog -- we cannot read from the commitlog (currently).

Memtable check is not deterministic, since the old data might be flushed to a new sstable already. So unless you do major, memtable check is not enough to prevent the problem. If we see *expired* tombstone is accumulating over time, that's not hard to determine based on metadata, then we have to piggyback on major. If we had to check row / cell of uncompacting sstables, that would be essentially running major, so why not do it?

Reply:
I think we can improve our tombstone compaction in ICS to escalate to system-wide flush + major if droppable ratio is growing beyond the configuration (tombstone_threshold). Note when we see reads are touching lots of tombstones, it may be out of compaction control if the tombstone creation rate is high, and tons of tombstones are created within the gc period. if repair tombstone mode is used, then we have to make sure repairs run more frequently, such that tombstones can be dropped. otherwise, the number of tombstones increase indefinitely, and reads start to suffer. the incremental repair will be really helpful to tackle this problem, since reducing the repair work in each iteration, allows us to increase repair frequency, therefore, keeping the amount of tombstones in the system low.

Reply:
> > Suggestion: purge_tombstones_in_commitlog_and_memtables
> 
> This is not accurate, what we want is commitlog and memtables impeding the tombstone GC process, preventing us from collecting tombstones because of (false) data shadowing.

Finding a descriptive name for it isn't easy. But aggressive_blah_blah isn't good. What if we find another way to improve tombstone garbage collection?

How about: consider_all_local_data.

(actually this name argues for it to be the default, since isn't major compaction supposed to consider all local data in the first place?)

> 
> > What are reasons not to use it (or: why aren't we making it the only option)? What are the tradeoffs?
> 
> There are 2 reasons AFAIK:
> 
> * Performance: this PR adds some more work to the major, although considering the cost of a major, this angle is perhaps not that significant

I guess for large tables the overhead is minor, for small tables it is significant.


> * Sheer paranoia. We are touching a very sensitive area here and although we are pretty sure about the change being correct, data resurrection is such a scary ghost that we want an escape hatch.

Let's keep the option to switch defaults. Make it a bool switch, so we can flip the default from false to true as we gain experience, and still retain the option of going back.